examples/state_machine/states.py

```diff
@@ -0,0 +1,361 @@
+        return PNTransition(state=HandshakeReconnectingState)
+
+
+class ReceivingState(PNState):
+    def __init__(self, context: PNContext) -> None:
+        super().__init__(context)
+        self._transitions = {
+            events.SubscriptionChangedEvent.__name__: self.subscription_changed,
+            events.SubscriptionRestoredEvent.__name__: self.subscription_restored,
+            events.ReceiveSuccessEvent.__name__: self.receiving_success,
+            events.ReceiveFailureEvent.__name__: self.receiving_failure,
+            events.DisconnectEvent.__name__: self.disconnect,
+            events.ReconnectEvent.__name__: self.reconnect,
+        }
+
+    def on_enter(self, context: Union[None, PNContext]):
+        super().on_enter(context)
+        return effects.ReceiveEventsEffect(context.channels, context.groups)
+
+    def on_exit(self):
+        super().on_exit()
+        return effects.CancelReceiveEventsEffect()
+
+    def subscription_changed(self, event: events.SubscriptionChangedEvent, context: PNContext) -> PNTransition:
+        return PNTransition(state=self.__class__, context=self._context)
+
+    def subscription_restored(self, event: events.SubscriptionRestoredEvent, context: PNContext) -> PNTransition:
+        return PNTransition(state=self.__class__, context=self._context)
+
+    def receiving_success(self, event: events.ReceiveSuccessEvent, context: PNContext) -> PNTransition:
+        return PNTransition(state=self.__class__, context=self._context)
+
+    def receiving_failure(self, event: events.ReceiveFailureEvent, context: PNContext) -> PNTransition:
+        return PNTransition(
+            state=ReceiveReconnectingStagte
+        )
+
+    def disconnect(self, event: events.DisconnectEvent, context: PNContext) -> PNTransition:
+        return PNTransition(state=ReceiveStoppedState)
+
+
+class ReceiveReconnectingStagte(PNState):
+    def __init__(self, context: PNContext) -> None:
+        super().__init__(context)
+        self._transitions = {
+
+        }
+
+
+class ReceiveFailedState(PNState):
+    def __init__(self, context: PNContext) -> None:
+        super().__init__(context)
+        self._transitions = {
+            events.ReceiveReconnectRetryEvent.__name__: self.reconnect_retry,
+            events.SubscriptionChangedEvent.__name__: self.subscription_changed,
+            events.SubscriptionRestoredEvent.__name__: self.subscription_restored,
+            events.ReconnectEvent.__name__: self.reconnect,
+        }
+
+    def reconnect_retry(self, event: events.HandshakeReconnectRetryEvent, context: PNContext) -> PNTransition:
+        return PNTransition(
+            state=ReceivingState,
+            context=self._context
+        )
+
+    def subscription_changed(self, event: events.SubscriptionChangedEvent, context: PNContext) -> PNTransition:
+        return PNTransition(
+            state=ReceivingState,
+            context=self._context
```

Comment:
Once again I'm a little concerned about the context here. Not everything will come with event, but some information will. In this case we'll have new `channels` and `channel_groups` yet it seems like nothing is going to be taken from the `context`, and everything will come from the current state `self._context`. I would expect a merge of both. Please review the context passing and make sure that for every event every piece of information coming with it is taken into the next state and only if something is missing we'll take it from `self._context`